OpenRA.Mods.RA/Activities/LayMines.cs

```diff
@@ -18,72 +20,93 @@
-					&& buildings.Contains(a.Info.Name))
-					.ClosestTo(self);
-
-				if (rearmTarget == null)
-					return new Wait(20);
-
-				return Util.SequenceActivities(
-					new MoveAdjacentTo(self, Target.FromActor(rearmTarget)),
-					movement.MoveTo(self.World.Map.CellContaining(rearmTarget.CenterPosition), rearmTarget),
-					new Rearm(self, info.RearmSound),
-					new Repair(rearmTarget),
-					this);
-			}
 
-			var ml = self.Trait<Minelayer>();
-			if (ml.Minefield.Contains(self.Location) && ShouldLayMine(self, self.Location))
+			if (ammoPools != null)
+				foreach (var pool in ammoPools)
+				{
+					if (pool.Info.Name == info.AmmoPool && !pool.HasAmmo())
+					{
+						// Rearm & repair at fix, then back out here to refill the minefield some more
+						var buildings = info.RearmBuildings;
+						var rearmTarget = self.World.Actors.Where(a => self.Owner.Stances[a.Owner] == Stance.Ally
+							&& buildings.Contains(a.Info.Name))
+							.ClosestTo(self);
+
+						if (rearmTarget == null)
+							return new Wait(20);
+
+						return Util.SequenceActivities(
+							new MoveAdjacentTo(self, Target.FromActor(rearmTarget)),
+							movement.MoveTo(self.World.Map.CellContaining(rearmTarget.CenterPosition), rearmTarget),
+							new Rearm(self),
+							new Repair(rearmTarget),
+							this);
+					}
+				}
+
+			if (minelayer.Minefield.Contains(self.Location) && ShouldLayMine(self, self.Location))
 			{
 				LayMine(self);
-				return Util.SequenceActivities(new Wait(20), this); // a little wait after placing each mine, for show
+				return Util.SequenceActivities(new Wait(20), this); // A little wait after placing each mine, for show
 			}
 
-			if (ml.Minefield.Length > 0)
+			if (minelayer.Minefield.Length > 0)
 			{
-				// dont get stuck forever here
+				// Don't get stuck forever here
 				for (var n = 0; n < 20; n++)
 				{
-					var p = ml.Minefield.Random(self.World.SharedRandom);
+					var p = minelayer.Minefield.Random(self.World.SharedRandom);
 					if (ShouldLayMine(self, p))
 						return Util.SequenceActivities(movement.MoveTo(p, 0), this);
 				}
 			}
 
-			// TODO: return somewhere likely to be safe (near fix) so we're not sitting out in the minefield.
+			// TODO: Return somewhere likely to be safe (near fix) so we're not sitting out in the minefield.
 			return new Wait(20);	// nothing to do here
 		}
 
 		static bool ShouldLayMine(Actor self, CPos p)
 		{
-			// if there is no unit (other than me) here, we want to place a mine here
+			// If there is no unit (other than me) here, we want to place a mine here
 			return !self.World.ActorMap.GetUnitsAt(p).Any(a => a != self);
 		}
 
 		static void LayMine(Actor self)
 		{
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null)
-				limitedAmmo.TakeAmmo();
+			var info = self.Info.Traits.Get<MinelayerInfo>();
+			var ammoPools = self.TraitsImplementing<AmmoPool>();
+
+			if (ammoPools != null)
```

Comment:
This `if`, with the `forach` and `if` it contains, can be simplified to
>			if (ammoPools != null)
			{
				var pool = ammoPools.FirstOrDefault(x => x.Info.Name == info.AmmoPool);
				if (pool != null)
					pool.TakeAmmo();
			}